staging/src/k8s.io/client-go/discovery/cached/disk/compactable_disk_cache.go

```diff
@@ -0,0 +1,236 @@
+			klog.Errorf("error touching file: %s", err)
+		}
+		// "singleflight" ensures that only one disk compaction occurs at a time within
+		// the current process.
+		var requestGroup singleflight.Group
+		begin := time.Now()
+		numFilesEvicted, err, _ := requestGroup.Do("disk-cache-compaction", func() (interface{}, error) {
+			return compactDiskCache(cacheDir, cacheSize)
+		})
+		finish := time.Now()
+		if err != nil {
+			klog.Errorf("Error during disk cache compaction: %s", err)
+		}
+		klog.V(7).Infof("%d files evicted from cache, compaction time: %s",
+			numFilesEvicted.(int), finish.Sub(begin))
+	}
+}
+
+// shouldCompactCache returns true if the disk cache compaction
+// trigger condition is met; false otherwise. The trigger condition
+// is if the modification time of the passed file is greater than one
+// day.
+func shouldCompactCache(fullFilePath string) bool {
+	fileInfo, err := os.Stat(fullFilePath)
+	if err != nil {
+		return false
+	}
+	durationSinceLastChange := time.Now().Sub(fileInfo.ModTime())
+	if durationSinceLastChange > oneDay {
+		return true
+	}
+	return false
+}
+
+// touchFile updates the passed file's access time and modification
+// time to the current time. Returns an error if one occurred while
+// updating the file.
+func touchFile(fullFilePath string) error {
+	currentTime := time.Now()
+	return os.Chtimes(fullFilePath, currentTime, currentTime)
+}
+
+// compactDiskCache removes old cache files (determined by modification time)
+// if the current cache size is larger than "cacheSize". Walks the entire
+// disk cache, and sorts files by ascending modification time. Removes files
+// until the disk cache is compacted below the "cacheSize" threshold. Returns
+// the number of files evicted from the cache and an error if one occurred.
+func compactDiskCache(cacheDir string, cacheSize int64) (int, error) {
+	currentCacheSize := int64(0)
+	files := []*fileInfo{}
+	// Walk the disk cache files to determine the cache size and file
+	// modification times.
+	err := filepath.Walk(cacheDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		fileSize := info.Size()
+		files = append(files, &fileInfo{
+			filePath: path,
+			modTime:  info.ModTime(),
+			fileSize: fileSize,
+		})
+		currentCacheSize += fileSize
+		return nil
+	})
+	if err != nil {
+		return 0, err
+	}
+	// Only compact disk cache if it is beyond the "cacheSize" threshold. This
+	// threshold is NOT a hard limit. The cache will breach this threshold, but
+	// the compaction will return the disk cache size below the "cacheSize".
+	klog.V(7).Infof("Disk cache current size: %d", currentCacheSize)
+	klog.V(7).Infof("Disk cache threshold size: %d", cacheSize)
+	numFilesEvicted := 0
+	if currentCacheSize > cacheSize {
+		klog.V(7).Infoln("Removing old cache files...")
+		// Starting with the oldest files (by modification time), delete each
```

Comment:
I would sure feel better about this if we touched files on access, do we?

Reply:
Yes. Every file that is retrieved with `Get` is touched.

Reply:
Ah, I did ask for that, sorry! 😳 

I suggest a different algorithm: put all files into a random order, and delete as many as needed in that order to get the size to the desired level. Since commonly needed files will be quickly added back and less common ones won't, I think this is OK.

Or at least, if we go by creation time, we could end up doing much worse?